### DISCUSSION TITLE
Weird Prisma output behavior

### DISCUSSION THREAD

**Discussion by Digital39999** [prisma/prisma#25127] [Q&A]
### Question

While trying to make a generator which would use Prisma's client type file, specifically (`node_modules/.prisma/client/index.d.ts` by default), i noticed there is weird behavior when changing the output directory manually, so:

- docs say default client folder is: `node_modules/.prisma/client`, but it generates in `node_modules/@prisma/client` with reference to 1st folder?
- but when explicitly overriding output folder to lets say `../client` (in root) or even `node_modules/@prisma/whatever`, it generates direct files, which don't reference back to .prisma folder, why?

how can i 100% accurately always get original `index.d.ts` file?
It's also not specified in `options.otherGenerators` of `generatorHandler.onGenerate` when using `@prisma/generator-helper`.

**Comment by Digital39999** [comment]
Currently this works out for me, but really?

![Ck7zae3zM](https://github.com/user-attachments/assets/450f13be-d05c-4690-ba60-d3ecc0e3cfe8)

**Comment by ludralph** [comment]
Hi @Digital39999 👋 

By default, when you run `prisma generate`, the Prisma Client is generated into the `./node_modules/.prisma/client` directory. This behavior ensures that the client is readily available for import using the `@prisma/client` package in your project.

When you specify a custom output directory in the generator block of your `schema.prisma` file, Prisma generates the client in the specified location. 

For example, with the following generator configuration:
```prisma
generator client {
  provider = "prisma-client-js"
  output   = "../src/generated/client"
}
```
To ensure that you are always accessing the original `index.d.ts` file, you should import the Prisma Client directly from the custom output directory. For instance:
```js
import { PrismaClient } from '../src/generated/client';
```

The `index.d.ts` file in node_modules/@prisma/client will still reference the default `.prisma/client directory`. To avoid confusion, always import the client from the custom output directory if specified.

**Reply by DevinJohw** [reply]
Thanks, you save my day.

**Comment by Digital39999** [comment]
Hi @ludralph 👋

Thank you for the clarification, but there seems to be an issue with this approach. If the package itself depends on `@prisma/client` and includes a Prisma schema for testing or other internal purposes, it will reference the generated client within its own environment rather than the project it's being used in. This can lead to confusion when trying to import `@prisma/client` in the project where the package is being used because it ends up pointing to the client generated within the package's `node_modules` rather than the one in the consuming project.

Another source of confusion is the inconsistency in behavior when no output directory is specified. When no output is provided, Prisma generates the client in `./node_modules/.prisma/client`, but it also generates a `node_modules/@prisma/client/default.d.ts` file that re-exports from that location. However, this does not happen when an explicit output is specified. This makes it unclear when to follow these re-exports and when not to, especially since this behavior isn't well-documented.

Additionally, it's difficult to know if the `default.d.ts` file was generated or if the `default` export path is being used when working across different package managers. This ambiguity makes it challenging to determine which `@prisma/client` version or path should be referenced in various contexts.

Without the ability to consistently import `@prisma/client` across different setups, it becomes tricky to manage dependencies, especially in environments where multiple packages might include their own Prisma schemas or rely on custom outputs. A clearer, more consistent approach would be beneficial in these cases to avoid potential issues with path resolution and client generation.

**Reply by ludralph** [reply]
Hi @Digital39999 

Thank you for getting back to me. Could you please these improvement as [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that our engineering team could take a look?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
